evap/staff/views.py

```diff
@@ -628,40 +650,58 @@ def course_participant_import(request, semester_id, course_id):
-        if excel_file:
-            imported_users = UserImporter.process(request, excel_file, test_run)
-        else:
-            imported_users = import_course.participants.all()
-
-        # Print message for test run.
-        if test_run and imported_users:
-            messages.success(request, "%d Participants would be added to course %s" % (len(imported_users), course.name))
+        if test_run:
+            delete_import_file(request.user.id)  # remove old file if still exists
+            excel_form.excel_file_required = True
+            if excel_form.is_valid():
+                excel_file = excel_form.cleaned_data['excel_file']
+                file_content = excel_file.read()
+                # if on a test run, the process method will not return a list of users that would be
+                # imported so there is currently no way to show how many users would be imported.
+                __, success_messages, warnings, errors = UserImporter.process(file_content, test_run)
+                if not errors:
+                    save_import_file(excel_file, request.user.id)
+
+        elif import_run:
+            file_content = get_import_file_content_if_exists(request.user.id)
+            if not file_content:
+                raise SuspiciousOperation("No test run performed previously.")
+
+            imported_users, success_messages, warnings, __ = UserImporter.process(file_content, test_run)
+            delete_import_file(request.user.id)
+            course.participants.add(*imported_users)
+            success_messages.append(("{} Participants added to course {}").format(len(imported_users), course.name))
+            forward_messages(request, success_messages, warnings)
+            return redirect('staff:semester_view', semester_id)
 
-        # Test run, or an error occurred while parsing -> stay and display error.
-        if test_run or not imported_users:
-            return render(request, "staff_course_participant_import.html", dict(course=course, form=form))
-        else:
-            # Add users to course participants. * converts list into parameters.
+        elif copy_run and copy_form.is_valid():
+            import_course = copy_form.cleaned_data['course']
+            imported_users = import_course.participants.all()
             course.participants.add(*imported_users)
             messages.success(request, "%d Participants added to course %s" % (len(imported_users), course.name))
             return redirect('staff:semester_view', semester_id)
-    else:
-        return render(request, "staff_course_participant_import.html", dict(course=course, form=form, semester=semester))
+
+    test_passed = import_file_exists(request.user.id)
+    return render(request, "staff_course_participant_import.html", dict(
+        course=course, excel_form=excel_form, copy_form=copy_form, semester=semester, success_messages=success_messages,
+        warnings=warnings, errors=errors, participant_test_passed=test_passed,))
```

Comment:
The trailing comma was intended (and missing on the other dicts, I admit :) ). I heard @janno42 support this style because the line won't be changed if you append another item in a new line (see http://stackoverflow.com/questions/11597901/why-does-python-allow-a-trailing-comma-in-list#11597911). Since the lines are already pretty full, the next item would probably be appended on the next line, so this would apply. What do you think?